### PR TITLE
Alter testing

### PR DESCRIPTION
**Pull request by rainwoodman**
Take a look.

I think this is easier to use than the current makefile. 
This works better with tab-completion.

### CODE REVIEW

**Comment by rainwoodman** [comment]
I added cmocka as a built-in dependency, which is only build if tests are requested.

The version on travis is too old and it conflicts with our usage. 

I also significantly shrinked the duration of the example travis run. We shall gradually move to rely on the unit tests instead of this final integration test.

**Comment by rainwoodman** [comment]
This will likely make statictree green, and forcetree-test will need a rebase (for some filename changes and c99 syntax fixes.)

File: Makefile.tests [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+.PHONY: run-tests build-tests
+
+
+depends/lib/libcmocka.a:
+	cd depends; make lib/libcmocka.a
+
+build-tests: $(TESTBIN)
+
+run-tests: build-tests
```
Comment (sbird) [inline Makefile.tests]: Can you make this just "make test" as before? That is a standard make target name.

File: Makefile.tests [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	for tt in $(SUITE) ; do $$tt ; done
+
+clean-tests:
+	rm -f $(TESTBIN)
+
+$(TESTBIN):
+
+tests/test_%: tests/test_%.c $(DESTDIR)/%.o $(DESTDIR)/endrun.o depends/lib/libcmocka.a
```
Comment (sbird) [inline Makefile.tests]: If you link to endrun.o you need to call MPI_Init in main and MPI_Finalize in teardown for all the test modules.

Comment (rainwoodman) [reply Makefile.tests]: Yes. we shall certainly do that -- they have to deal with MPI at some point.

**Comment by rainwoodman** [comment]
What are the gains using the installed version of cmocka?

File: Makefile.tests (hunk 1 above)
Comment (rainwoodman) [reply Makefile.tests]: We use `run-tests.py` in many python packages -- that's where the name is from. I guess make is already a verb then it doesn't make sense to have two verbs.

**Comment by sbird** [comment]
You forgot to git add tests/stub.c

**Comment by sbird** [comment]
I guess the main practical advantage is that we don't have to compile cmocka ourselves - and thus we don't need cmake.

**Comment by rainwoodman** [comment]
In nbodykit, dealing with testing on a variety of number of ranks we use MPI_Comm_split and pass in the communicator to the testee, calling it with each different requested MPI size.

In MP_Gadget, the issue is our testee assumes they work on MPI_COMM_WORLD. It is a pretty poor design decision inherited from Gadget. but it will be too much work to fix that.

**Comment by rainwoodman** [comment]
If you have the luxury to install a binary of cmocka, then probably have the same luxury to install cmake?

The advantage of bundling a cmocka is that we require a somewhat recent cmocka (>1.0?). On travis the default cmocka is 0.3.x and we fail compiling the tests.

I made it such that if the tests are not requested cmocka is not built. Therefore it doesn't get into the way of a quick-and-dirty user, assuming tests are only meaningful for developers.

**Comment by sbird** [comment]
Sure sure - but if the issue is travis, why not just build it in the travis startup script? But it's not a big deal.

**Comment by rainwoodman** [comment]
Similar issue may happen on other boxes. I assume the OS on travis is 'typical'. 
Building cmocka doesn't really take much time so it is not a big deal I agree.

**Comment by sbird** [comment]
Ok, sure. It is not too bad. I am happy with this, merge when you're ready.

**Pull request merged**